File: backend/src/modules/attendance/webAttendanceGate.ts
/**
 * Web attendance is allowed only on genuine iOS Safari.
 * Native Android/iOS app clients are unchanged.
 */

export type WebAttendanceClientKind = 'native_app' | 'ios_safari' | 'blocked_web';

export type WebAttendanceGateResult = {
  kind: WebAttendanceClientKind;
  allowed: boolean;
  message?: string;
  deviceLabel: string;
  browserLabel: string;
  userAgent: string;
  isIos: boolean;
  isSafari: boolean;
};

function isBrowserUserAgent(ua: string): boolean {
  return /Mozilla\/\d/i.test(ua) && /(Chrome|CriOS|Safari|Firefox|FxiOS|Edg|EdgiOS)\//i.test(ua);
}

/** True Safari on iOS — excludes Chrome/Firefox/Edge/Opera iOS wrappers. */
export function isIosSafariUserAgent(ua: string): boolean {
  if (!ua) return false;
  const isIos = /iPhone|iPad|iPod/i.test(ua);
  if (!isIos) return false;
  // Other browsers on iOS still include "Safari" in UA — exclude them first.
  if (/CriOS|FxiOS|EdgiOS|OPiOS|YaBrowser|DuckDuckGo/i.test(ua)) return false;
  // Desktop Chrome-style tokens shouldn't appear on genuine mobile Safari.
  if (/Chrome\//i.test(ua)) return false;
  return /Safari\//i.test(ua) && /Version\//i.test(ua);
}

export function browserLabelFromUa(ua: string): string {
  if (/CriOS/i.test(ua)) return 'Chrome (iOS)';
  if (/FxiOS/i.test(ua)) return 'Firefox (iOS)';
  if (/EdgiOS/i.test(ua)) return 'Edge (iOS)';
  if (/OPiOS/i.test(ua)) return 'Opera (iOS)';
  if (isIosSafariUserAgent(ua)) return 'Safari (iOS)';
  if (/Edg\//i.test(ua)) return 'Edge';
  if (/Chrome\//i.test(ua) && /Safari\//i.test(ua)) return 'Chrome';
  if (/Firefox\//i.test(ua)) return 'Firefox';
  if (/Safari\//i.test(ua) && /Version\//i.test(ua)) return 'Safari';
  return 'Unknown browser';
}

export function deviceLabelFromUa(ua: string): string {
  if (/iPhone/i.test(ua)) return 'iPhone';
  if (/iPad/i.test(ua)) return 'iPad';
  if (/iPod/i.test(ua)) return 'iPod';
  if (/Android/i.test(ua)) return 'Android device';
  if (/Windows/i.test(ua)) return 'Windows PC';
  if (/Macintosh|Mac OS X/i.test(ua)) return 'Mac';
  if (/Linux/i.test(ua)) return 'Linux PC';
  return 'Unknown device';
}

/**
 * Decide whether this punch/register request may use the web attendance path.
 * Native app (android/ios platform + non-browser UA) always allowed.
 */
export function evaluateWebAttendanceGate(opts: {
  userAgent?: string | null;
  deviceInfo?: Record<string, unknown> | null;
}): WebAttendanceGateResult {
  const userAgent = String(opts.userAgent ?? '').trim();
  const platform = String(opts.deviceInfo?.platform ?? '').toLowerCase();
  const isIos = /iPhone|iPad|iPod/i.test(userAgent);
  const isSafari = isIosSafariUserAgent(userAgent);

  // Native Flutter app: platform android/ios and UA is not a browser.
  if ((platform === 'android' || platform === 'ios') && !isBrowserUserAgent(userAgent)) {
    const deviceLabel =
      String(opts.deviceInfo?.model ?? opts.deviceInfo?.name ?? platform) || platform;
    return {
      kind: 'native_app',
      allowed: true,
      deviceLabel,
      browserLabel: 'Native app',
      userAgent,
      isIos: platform === 'ios',
      isSafari: false,
    };
  }

  // Register/punch from native app without deviceInfo still sends a Dart/IO user-agent.
  if (!isBrowserUserAgent(userAgent)) {
    return {
      kind: 'native_app',
      allowed: true,
      deviceLabel: platform || 'Native app',
      browserLabel: 'Native app',
      userAgent,
      isIos: platform === 'ios',
      isSafari: false,
    };
  }

  // Treat as web (including spoofed platform with browser UA).
  const deviceLabel = deviceLabelFromUa(userAgent);
  const browserLabel = browserLabelFromUa(userAgent);

  if (isIosSafariUserAgent(userAgent)) {
    return {
      kind: 'ios_safari',
      allowed: true,
      deviceLabel,
      browserLabel: 'Safari (iOS)',
      userAgent,
      isIos: true,
      isSafari: true,
    };
  }

  let message =
    'Web attendance is only allowed on Safari on an iPhone or iPad. Use the mobile app on other devices.';
  if (isIos && !isSafari) {
    message =
      'On iPhone/iPad, attendance from the browser is only allowed in Safari. Chrome and other browsers are blocked — open this site in Safari, or use the mobile app.';
  } else if (!isIos && /Safari\//i.test(userAgent)) {
    message =
      'Safari on non-iOS devices cannot be used for attendance. Use the mobile app, or Safari on an iPhone/iPad.';
  }

  return {
    kind: 'blocked_web',
    allowed: false,
    message,
    deviceLabel,
    browserLabel,
    userAgent,
    isIos,
    isSafari: false,
  };
}

File: nb_crm_flutter/lib/features/attendance/presentation/web_attendance_gate.dart
import 'package:device_info_plus/device_info_plus.dart';
import 'package:flutter/foundation.dart';

/// Web attendance is allowed only on genuine Safari running on an iPhone/iPad.
/// Desktop Safari, Android browsers, and Chrome/Firefox/Edge on iOS are blocked.
/// Native mobile apps never use this gate (`kIsWeb` is false).
class WebAttendanceGate {
  WebAttendanceGate._();

  static Future<WebAttendanceStatus>? _cached;

  static Future<WebAttendanceStatus> evaluate() {
    _cached ??= _evaluate();
    return _cached!;
  }

  /// Force re-check (e.g. after Refresh Status).
  static Future<WebAttendanceStatus> reevaluate() {
    _cached = _evaluate();
    return _cached!;
  }

  static Future<WebAttendanceStatus> _evaluate() async {
    if (!kIsWeb) {
      return const WebAttendanceStatus(
        allowed: true,
        isWeb: false,
        isIos: false,
        isSafari: false,
        deviceLabel: 'Native app',
        browserLabel: 'App',
      );
    }

    try {
      final web = await DeviceInfoPlugin().webBrowserInfo;
      final ua = web.userAgent ?? '';
      final uaLower = ua.toLowerCase();
      final isIos = uaLower.contains('iphone') ||
          uaLower.contains('ipad') ||
          uaLower.contains('ipod');
      final isOtherIosBrowser = uaLower.contains('crios') ||
          uaLower.contains('fxios') ||
          uaLower.contains('edgios') ||
          uaLower.contains('opios') ||
          uaLower.contains('yabrowser') ||
          uaLower.contains('duckduckgo');
      // Chrome desktop also embeds "Safari/" — require Version/ and no Chrome/.
      final hasChrome = uaLower.contains('chrome/');
      final isSafari = !isOtherIosBrowser &&
          !hasChrome &&
          uaLower.contains('safari/') &&
          uaLower.contains('version/') &&
          (web.browserName == BrowserName.safari || isIos);

      final deviceLabel = _deviceLabel(uaLower);
      final browserLabel = _browserLabel(
        uaLower: uaLower,
        isIos: isIos,
        isSafari: isSafari && isIos,
        isOtherIosBrowser: isOtherIosBrowser,
      );

      if (isIos && isSafari) {
        return WebAttendanceStatus(
          allowed: true,
          isWeb: true,
          isIos: true,
          isSafari: true,
          deviceLabel: deviceLabel,
          browserLabel: 'Safari (iOS)',
          userAgent: ua,
        );
      }

      String message;
      if (isIos && !isSafari) {
        message =
            'On iPhone/iPad, browser attendance works only in Safari. '
            'Chrome and other browsers are blocked. Open this site in Safari, or use the mobile app.';
      } else if (!isIos && uaLower.contains('safari/')) {
        message =
            'Safari on non-iOS devices cannot be used for attendance. '
            'Use the mobile app, or Safari on an iPhone/iPad.';
      } else {
        message =
            'Web attendance is disabled on this browser. '
            'Use the mobile app, or Safari on an iPhone/iPad only.';
      }

      return WebAttendanceStatus(
        allowed: false,
        isWeb: true,
        isIos: isIos,
        isSafari: false,
        deviceLabel: deviceLabel,
        browserLabel: browserLabel,
        userAgent: ua,
        message: message,
      );
    } catch (_) {
      return const WebAttendanceStatus(
        allowed: false,
        isWeb: true,
        isIos: false,
        isSafari: false,
        deviceLabel: 'Unknown',
        browserLabel: 'Unknown',
        message:
            'Unable to verify this browser. Web attendance is only allowed in Safari on iPhone/iPad.',
      );
    }
  }

  static String _deviceLabel(String uaLower) {
    if (uaLower.contains('iphone')) return 'iPhone';
    if (uaLower.contains('ipad')) return 'iPad';
    if (uaLower.contains('ipod')) return 'iPod';
    if (uaLower.contains('android')) return 'Android device';
    if (uaLower.contains('windows')) return 'Windows PC';
    if (uaLower.contains('mac os') || uaLower.contains('macintosh')) return 'Mac';
    if (uaLower.contains('linux')) return 'Linux PC';
    return 'Unknown device';
  }

  static String _browserLabel({
    required String uaLower,
    required bool isIos,
    required bool isSafari,
    required bool isOtherIosBrowser,
  }) {
    if (isSafari) return 'Safari (iOS)';
    if (uaLower.contains('crios')) return 'Chrome (iOS)';
    if (uaLower.contains('fxios')) return 'Firefox (iOS)';
    if (uaLower.contains('edgios')) return 'Edge (iOS)';
    if (isOtherIosBrowser) return 'Other browser (iOS)';
    if (uaLower.contains('edg/')) return 'Edge';
    if (uaLower.contains('chrome/')) return 'Chrome';
    if (uaLower.contains('firefox/')) return 'Firefox';
    if (uaLower.contains('safari/') && uaLower.contains('version/')) {
      return isIos ? 'Safari' : 'Safari (desktop)';
    }
    return 'Unknown browser';
  }
}

class WebAttendanceStatus {
  const WebAttendanceStatus({
    required this.allowed,
    required this.isWeb,
    required this.isIos,
    required this.isSafari,
    required this.deviceLabel,
    required this.browserLabel,
    this.userAgent,
    this.message,
  });

  final bool allowed;
  final bool isWeb;
  final bool isIos;
  final bool isSafari;
  final String deviceLabel;
  final String browserLabel;
  final String? userAgent;
  final String? message;
}
